File: financial-statement-studio/apps/web/src/types/ledger-account.ts
export const ACCOUNT_TYPE_OPTIONS = [
  {
    value: "asset",
    label: "Assets",
  },
  {
    value: "liability",
    label: "Liabilities",
  },
  {
    value: "equity",
    label: "Equity",
  },
  {
    value: "revenue",
    label: "Revenue and income",
  },
  {
    value: "expense",
    label: "Expenses",
  },
] as const;

export const REPORT_CATEGORY_OPTIONS = [
  {
    value: "revenue",
    label: "Revenue",
  },
  {
    value: "cost_of_sales",
    label: "Cost of sales",
  },
  {
    value: "direct_service_costs",
    label: "Direct service costs",
  },
  {
    value: "manufacturing_costs",
    label: "Manufacturing costs",
  },
  {
    value: "other_income",
    label: "Other income",
  },
  {
    value: "administrative_expenses",
    label: "Administrative expenses",
  },
  {
    value: "selling_distribution_expenses",
    label: "Selling and distribution expenses",
  },
  {
    value: "finance_costs",
    label: "Finance costs",
  },
  {
    value: "taxation",
    label: "Taxation",
  },
  {
    value: "current_assets",
    label: "Current assets",
  },
  {
    value: "non_current_assets",
    label: "Non-current assets",
  },
  {
    value: "current_liabilities",
    label: "Current liabilities",
  },
  {
    value: "non_current_liabilities",
    label: "Non-current liabilities",
  },
  {
    value: "equity",
    label: "Equity",
  },
] as const;

export const CASH_FLOW_CATEGORY_OPTIONS = [
  {
    value: "operating",
    label: "Operating activity",
  },
  {
    value: "investing",
    label: "Investing activity",
  },
  {
    value: "financing",
    label: "Financing activity",
  },
  {
    value: "non_cash",
    label: "Non-cash item",
  },
  {
    value: "not_applicable",
    label: "Not applicable",
  },
] as const;

export const NORMAL_BALANCE_OPTIONS = [
  {
    value: "debit",
    label: "Debit",
  },
  {
    value: "credit",
    label: "Credit",
  },
] as const;

export type AccountType =
  (typeof ACCOUNT_TYPE_OPTIONS)[number]["value"];

export type ReportCategory =
  (typeof REPORT_CATEGORY_OPTIONS)[number]["value"];

export type CashFlowCategory =
  (typeof CASH_FLOW_CATEGORY_OPTIONS)[number]["value"];

export type NormalBalance =
  (typeof NORMAL_BALANCE_OPTIONS)[number]["value"];

export const REPORT_CATEGORY_ACCOUNT_TYPE: Record<
  ReportCategory,
  AccountType
> = {
  revenue: "revenue",
  other_income: "revenue",

  cost_of_sales: "expense",
  direct_service_costs: "expense",
  manufacturing_costs: "expense",
  administrative_expenses: "expense",
  selling_distribution_expenses: "expense",
  finance_costs: "expense",
  taxation: "expense",

  current_assets: "asset",
  non_current_assets: "asset",

  current_liabilities: "liability",
  non_current_liabilities: "liability",

  equity: "equity",
};

export const ACCOUNT_TYPE_DEFAULT_NORMAL_BALANCE: Record<
  AccountType,
  NormalBalance
> = {
  asset: "debit",
  expense: "debit",
  liability: "credit",
  equity: "credit",
  revenue: "credit",
};

export type LedgerAccount = {
  id: string;
  company_id: string;
  parent_account_id: string | null;

  account_code: string;
  account_name: string;

  account_type: AccountType;
  report_category: ReportCategory;
  cash_flow_category: CashFlowCategory | null;
  normal_balance: NormalBalance;

  description: string | null;

  is_system_account: boolean;
  is_cash_equivalent: boolean;
  is_active: boolean;
  display_order: number;

  created_at: string;
  updated_at: string;
};

export type LedgerAccountListResponse = {
  items: LedgerAccount[];
  total: number;
  offset: number;
  limit: number;
};

export type ChartInitializationResponse = {
  company_id: string;
  business_template: string;
  created_count: number;
  skipped_count: number;
  items: LedgerAccount[];
};

export type LedgerAccountCreatePayload = {
  account_code: string;
  account_name: string;
  account_type: AccountType;
  report_category: ReportCategory;
  cash_flow_category?: CashFlowCategory | null;
  is_cash_equivalent?: boolean;
  normal_balance: NormalBalance;
  parent_account_id?: string | null;
  description?: string | null;
  display_order: number;
};

export type LedgerAccountUpdatePayload = Partial<
  LedgerAccountCreatePayload
>;

export type LedgerAccountListQuery = {
  search?: string;
  accountType?: AccountType;
  reportCategory?: ReportCategory;
  includeInactive?: boolean;
  offset?: number;
  limit?: number;
};

File: financial-statement-studio/apps/api/app/schemas/ledger_account.py
from datetime import datetime
from enum import StrEnum
from typing import Any
from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
)


class AccountType(StrEnum):
    """Main accounting classifications."""

    ASSET = "asset"
    LIABILITY = "liability"
    EQUITY = "equity"
    REVENUE = "revenue"
    EXPENSE = "expense"


class ReportCategory(StrEnum):
    """Financial-statement sections used by the reporting engine."""

    REVENUE = "revenue"
    COST_OF_SALES = "cost_of_sales"
    DIRECT_SERVICE_COSTS = "direct_service_costs"
    MANUFACTURING_COSTS = "manufacturing_costs"
    OTHER_INCOME = "other_income"

    ADMINISTRATIVE_EXPENSES = (
        "administrative_expenses"
    )
    SELLING_DISTRIBUTION_EXPENSES = (
        "selling_distribution_expenses"
    )
    FINANCE_COSTS = "finance_costs"
    TAXATION = "taxation"

    CURRENT_ASSETS = "current_assets"
    NON_CURRENT_ASSETS = "non_current_assets"
    CURRENT_LIABILITIES = "current_liabilities"
    NON_CURRENT_LIABILITIES = (
        "non_current_liabilities"
    )
    EQUITY = "equity"


class CashFlowCategory(StrEnum):
    """Cash-flow classification for an account."""

    OPERATING = "operating"
    INVESTING = "investing"
    FINANCING = "financing"
    NON_CASH = "non_cash"
    NOT_APPLICABLE = "not_applicable"


class NormalBalance(StrEnum):
    """The side on which an account normally increases."""

    DEBIT = "debit"
    CREDIT = "credit"


def clean_required_text(value: str) -> str:
    cleaned_value = " ".join(value.split())

    if not cleaned_value:
        raise ValueError(
            "This field cannot be empty.",
        )

    return cleaned_value


def clean_account_code(value: str) -> str:
    cleaned_value = value.strip().upper()

    if not cleaned_value:
        raise ValueError(
            "Account code cannot be empty.",
        )

    return cleaned_value


def clean_optional_text(value: Any) -> Any:
    if not isinstance(value, str):
        return value

    cleaned_value = value.strip()

    return cleaned_value or None


class LedgerAccountCreate(BaseModel):
    """Payload for creating a custom ledger account."""

    account_code: str = Field(
        min_length=1,
        max_length=30,
    )

    account_name: str = Field(
        min_length=2,
        max_length=180,
    )

    account_type: AccountType
    report_category: ReportCategory

    cash_flow_category: CashFlowCategory | None = None

    normal_balance: NormalBalance

    parent_account_id: str | None = Field(
        default=None,
        min_length=36,
        max_length=36,
    )

    description: str | None = Field(
        default=None,
        max_length=2000,
    )

    display_order: int = Field(
        default=0,
        ge=0,
        le=100000,
    )

    model_config = ConfigDict(
        str_strip_whitespace=True,
    )
    is_cash_equivalent: bool = False

    @field_validator("account_code")
    @classmethod
    def validate_account_code(
        cls,
        value: str,
    ) -> str:
        return clean_account_code(value)

    @field_validator("account_name")
    @classmethod
    def validate_account_name(
        cls,
        value: str,
    ) -> str:
        return clean_required_text(value)

    @field_validator(
        "parent_account_id",
        "description",
        mode="before",
    )
    @classmethod
    def validate_optional_text(
        cls,
        value: Any,
    ) -> Any:
        return clean_optional_text(value)


class LedgerAccountUpdate(BaseModel):
    """Payload for partially updating a ledger account."""

    account_code: str | None = Field(
        default=None,
        min_length=1,
        max_length=30,
    )

    account_name: str | None = Field(
        default=None,
        min_length=2,
        max_length=180,
    )

    account_type: AccountType | None = None
    report_category: ReportCategory | None = None

    cash_flow_category: CashFlowCategory | None = None

    normal_balance: NormalBalance | None = None

    parent_account_id: str | None = Field(
        default=None,
        min_length=36,
        max_length=36,
    )

    description: str | None = Field(
        default=None,
        max_length=2000,
    )

    display_order: int | None = Field(
        default=None,
        ge=0,
        le=100000,
    )

    model_config = ConfigDict(
        str_strip_whitespace=True,
    )
    is_cash_equivalent: bool | None = None

    @field_validator("account_code")
    @classmethod
    def validate_account_code(
        cls,
        value: str | None,
    ) -> str | None:
        if value is None:
            return None

        return clean_account_code(value)

    @field_validator("account_name")
    @classmethod
    def validate_account_name(
        cls,
        value: str | None,
    ) -> str | None:
        if value is None:
            return None

        return clean_required_text(value)

    @field_validator(
        "parent_account_id",
        "description",
        mode="before",
    )
    @classmethod
    def validate_optional_text(
        cls,
        value: Any,
    ) -> Any:
        return clean_optional_text(value)


class LedgerAccountResponse(BaseModel):
    """Ledger account returned by the API."""

    id: str
    company_id: str
    parent_account_id: str | None

    account_code: str
    account_name: str

    account_type: AccountType
    report_category: ReportCategory
    cash_flow_category: CashFlowCategory | None
    normal_balance: NormalBalance

    description: str | None

    is_system_account: bool
    is_cash_equivalent: bool
    is_active: bool
    display_order: int

    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


class LedgerAccountListResponse(BaseModel):
    """Paginated Chart of Accounts response."""

    items: list[LedgerAccountResponse]
    total: int
    offset: int
    limit: int


class ChartInitializationResponse(BaseModel):
    """Result of initializing a company's default accounts."""

    company_id: str
    business_template: str
    created_count: int
    skipped_count: int
    items: list[LedgerAccountResponse]